src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/RetryLoop.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Protocol.Core.v3;
+
+namespace NuGet.Protocol
+{
+    public class RetryLoop
+    {
+        private readonly int _maxTries;
+        private readonly TimeSpan _requestTimeout;
+        private readonly TimeSpan _retryDelay;
+
+        /// <summary>
+        /// The <see cref="RetryLoop"/> is for retrying and HTTP request if it times out, has any exception,
+        /// or returns a status code of 500 or greater.
+        /// </summary>
+        /// <param name="maxTries">
+        /// The maximum number of times to try the request. This value includes the initial attempt.
+        /// </param>
+        /// <param name="requestTimeout">How long to wait on the request to come back with a response.</param>
+        /// <param name="retryDelay">How long to wait before trying again after a failed request.</param>
+        public RetryLoop(int maxTries, TimeSpan requestTimeout, TimeSpan retryDelay)
+        {
+            _maxTries = maxTries;
+            _requestTimeout = requestTimeout;
+            _retryDelay = retryDelay;
+        }
+
+        public async Task<HttpResponseMessage> SendAsync(
+            HttpClient client,
+            HttpRequestMessage request,
+            HttpCompletionOption completionOption,
+            CancellationToken cancellationToken)
+        {
+            var tries = 0;
+            HttpResponseMessage response = null;
+            var success = false;
+
+            while (tries < _maxTries && !success)
+            {
+                if (tries > 0)
+                {
+                    request = request.Clone();
+                    await Task.Delay(_retryDelay, cancellationToken);
+                }
+
+                tries++;
+                success = true;
+
+                try
+                {
+                    // implement timeout
+                    using (var timeoutTcs = new CancellationTokenSource())
+                    using (var responseTcs = new CancellationTokenSource())
+                    using (cancellationToken.Register(() => responseTcs.Cancel()))
+                    {
+                        var timeoutTask = Task.Delay(_requestTimeout, timeoutTcs.Token);
+                        var responseTask = client.SendAsync(request, completionOption, responseTcs.Token);
+
+                        if (timeoutTask == await Task.WhenAny(responseTask, timeoutTask))
```

Comment:
some comments to explain the cleverness going here would make this more clear :smile: